pkg/coscheduling/coscheduling.go

```diff
@@ -95,294 +83,364 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
+			return framework.NewStatus(framework.Success, ""), 0
 
-	return framework.NewStatus(framework.Success, ""), 0
+		}
+		klog.Infof("cs.pgMgr.Permit error %v", err)
+		return framework.NewStatus(framework.Unschedulable, err.Error()), waitTime
+	}
+	if pgs = cs.pgMgr.PodGroupStatusCache().Get(fullName); pgs == nil {
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, "PodGroup not found"), waitTime
+	}
+	if wait := util.GetWaitTimeDuration(pgs.PodGroup, cs.scheduleTimeout); wait != 0 {
+		waitTime = wait
+	}
+	klog.V(5).Infof("Pod requires pgName %v", fullName)
+	if ready {
+		go cs.sendStartScheduleSignal(fullName)
+	}
+	return framework.NewStatus(framework.Wait, ""), waitTime
 }
 
-func (cs *Coscheduling) Reserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) *framework.Status {
+func (cs *Coscheduling) Reserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) *framework.Status {
 	return nil
 }
 
 // Unreserve rejects all other Pods in the PodGroup when one of the pods in the group times out.
 func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) {
-	pgInfo, _ := cs.getOrCreatePodGroupInfo(pod, time.Now())
-	if len(pgInfo.key) == 0 {
+	_, pgs, _ := cs.pgMgr.GetPodGroup(pod)
+	if pgs == nil {
 		return
 	}
-	podGroupName := pgInfo.name
-	cs.frameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
-		if waitingPod.GetPod().Namespace == pod.Namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
-			klog.V(3).Infof("Unreserve rejects the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
-			waitingPod.Reject(cs.Name())
+	pgs.AssumedPods.Delete(core.GetPodName(pod))
+}
+
+// PostBind is called after a pod is successfully bound. These plugins are used update PodGroup when pod is bound.
+func (cs *Coscheduling) PostBind(ctx context.Context, _ *framework.CycleState, pod *v1.Pod, nodeName string) {
+	klog.V(5).Infof("PostBind pod: %v/%v", pod.Namespace, pod.Name)
+	cs.pgMgr.PostBind(ctx, pod, nodeName)
+}
+
+// VerifyLocalCache verify pgstatus cache
+func (cs *Coscheduling) VerifyLocalCache(toSchedulePGName string) {
+	pgs := cs.pgMgr.PodGroupStatusCache().Get(toSchedulePGName)
+	// update pods cache
+	cs.frameworkHandler.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		klog.V(5).Info("Start IterateOverWaitingPods")
+		pod := waitingPod.GetPod()
+		pgName := util.VerifyPodLabelSatisfied(pod)
+		klog.V(5).Infof("Start walking through pod %v/%v uid: %v", pod.Namespace, pod.Name, pod.UID)
+		if pgName == toSchedulePGName {
+			klog.V(5).Infof("Wanted cache pod %v/%v", pod.Namespace, pod.Name)
+			_, found := pgs.AssumedPods.Get(core.GetPodName(pod))
+			if !found {
+				return
+			}
+			klog.V(6).Infof("Current pod group state %+v", pgs.AssumedPods.Items())
+			cs.pgMgr.PodGroupStatusCache().Set(util.GetPodGroupFullName(pgs.PodGroup), pgs)
```

Comment:
IIUC, these 2 lines will be called for all pods that belongs to "toSchedulePGName", right? I don't quite understand why we need to run Set() N times, and log the items N items.